src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/RazorSemanticTokensInfoService.cs

```diff
@@ -175,6 +184,71 @@ internal class RazorSemanticTokensInfoService : IRazorSemanticTokensInfoService
         }
 
         return razorRanges;
+
+        async Task<List<SemanticRange>?> GetTrimmedRangesForSemanticTokensAsync(
+            RazorCodeDocument codeDocument,
+            TextDocumentIdentifier textDocumentIdentifier,
+            Range razorRange,
+            RazorSemanticTokensLegend razorSemanticTokensLegend,
+            long documentVersion,
+            Guid correlationId,
+            CancellationToken cancellationToken)
+        {
+            var generatedDocument = codeDocument.GetCSharpDocument();
+            if (!TryGetCSharpRanges(codeDocument, razorRange, out var csharpRanges))
+            {
+                // There's no C# in the range.
+                return new List<SemanticRange>();
+            }
+
+            var csharpResponse = await GetMatchingCSharpResponseAsync(textDocumentIdentifier, documentVersion, csharpRanges, correlationId, cancellationToken).ConfigureAwait(false);
+
+            // Indicates an issue with retrieving the C# response (e.g. no response or C# is out of sync with us).
+            // Unrecoverable, return default to indicate no change. We've already queued up a refresh request in
+            // `GetMatchingCSharpResponseAsync` that will cause us to retry in a bit.
+            if (csharpResponse is null)
+            {
+                _logger.LogWarning("Issue with retrieving C# response for Razor range: ({startLine},{startChar})-({endLine},{endChar})", razorRange.Start.Line, razorRange.Start.Character, razorRange.End.Line, razorRange.End.Character);
+                return null;
+            }
+
+            var razorRanges = new List<SemanticRange>();
+            var colorBackground = _razorLSPOptionsMonitor.CurrentValue.ColorBackground;
+            var textClassification = razorSemanticTokensLegend.MarkupTextLiteral;
+            var razorSource = codeDocument.GetSourceText();
+
+            SemanticRange? previousSemanticRange = null;
+            Range? previousRazorSemanticRange = null;
+            for (var i = 0; i < csharpResponse.Length; i += TokenSize)
```

Comment:
This looks like its duplicated from the `GetCSharpSemanticRangesAsync` method. Can we share the code?